### PR TITLE
Padmiss support

### PR DESCRIPTION
**Pull request by electromuis**
With thanks to the finnish ITG community

Add support for folder based MemoryCard usage
Add support fot the https://padmiss.com/ network by saving scores to individual files.

(For communication with padmiss, use https://github.com/electromuis/padmiss-daemon)

### CODE REVIEW

**Comment by dguzek** [comment]
Thanks for this pull request!

I'm completely unfamiliar with padmiss, and the website didn't seem to offer any front-facing information about how it is used.  Additionally, padmiss-daemon doens't have a readme or anything similar from which one new to the project might glean what it does or how to use it.

Can you perhaps write a little bit more about this pull request?  What problem does it solve?  How does it accomplish that?  If merged, how would it be used by StepMania players?

**Comment by electromuis** [comment]
Its an online scoretracker which has been used in the past few ITG Eurocop events.
The flow works like:
- Cab admin a padmiss token (one to one case by now, still in development)
- Cab admin updates stepmania, and sets these lines:
MemoryCardOsMountPointP1=/media/dummy1 (or whatever)
MemoryCardOsMountPointP2=/media/dummy2 (or whatever)
MemoryCardDriver=Directory
MemoryCardPadmissEnabled=1
- Cab admin installs the daemon and configures it with the scanners
 - Daemon hooks into nfc reader modules (Search "rfid reader usb" on aliexpress/ebay) which normally act like a keyboard
- Player get a rfid card, depending on what scanner they use
- They register it over at https://padmiss.com/register
 - Player scans their card when you would normally instert your usb
 - The daemon reads the code, finds the player via de padmiss api. Then it makes a folder where stepmania expects the MemoryCard with all user relevant user data
 - Players plays a song
 - The StatsManaget saves an entry with the song/score/player data to a file in Save/Padmiss
 - The daemon notices a change in that folder,
 - The daemon reads the new file, and submits it to the api
 - The players opens the website or app and notices their new score

The daemon still needs an easily installable version, with that some documentation on configuration will come as well

**Comment by electromuis** [comment]
Windows build including these changes: https://drive.google.com/open?id=1_VFzXZg-fO4fxmYPV-k0ZCIalmlXM2LX

**Comment by Horsey-** [comment]
So when you say "app" can the app have NFC-enabled phones broadcast your RFID "card" to the reader? YEARS AGO there was a few apple engineers from the U of A that made a hook for Apple Wallet to display a login code for the same purpose (this never took off for whatever reason, mostly because at the time it was much more difficult to get your own private machine).

**Comment by electromuis** [comment]
With app i mean: https://github.com/electromuis/padmiss-app
Which is just another interface for the website.

@horsey i suppose you mean using your phone as your card? Depending on the scanners that could work.
Padmiss hasn't really settled on a standard (which might not even happen). For now we just use the cheapest option

**Comment by Alhetus** [comment]
Hi,

Some clarifications:
- the card readers have nothing to do with nfc (for now), they are the most basic usb rfid readers you can find on eBay, because 3$ is a really nice price point
- at this point Padmiss is pretty much just an API with a very basic and hacky react front, but the target is to improve the front greatly in the future to host tournaments and track global scores with player levels, achievements etc.
- anyone can basically create any player data fetching method for Padmiss if you just know the player rfidUid (a.k.a "secret", that can be any string really) that you post to the API.
- a "cab" is currently in charge of posting the scores to the API with a designated API key, I have future plans to actually have the player log in more securely (credentials or rfid card + pin code) and post the scores with a player token instead

**Comment by coderjo** [comment]
I have a question. first, do I have the sequence of operations correct here?

Upon scanning a card and successfully retrieving the info from the padmiss API, it creates the profile directory and populates it with a very basic stats.xml and editable.ini.

The MCDT_Folder class notices it in the memory card worker thread polling and tells MCM there is a card there, which the game loads.

Then the game, after each stage, writes stats in a special subdirectory of the profile, which the daemon notices and sends to padmiss.

After the game ends, how does the profile directory get removed? It appears that it must be at some point, because the MCDT_Folder code only appears to check if the profile folder exists, not that the profile data has changed.

I'm guessing this was only used on linux and/or macos so far, and not windows, correct? I think this might need minor tweaks for windows (the stat function is named _stat on windows, as is the struct).

Oh, and you should probably change the copyright on MemoryCardDriverThreaded_Folder.*, since it's your code. You just started with a class structure to implement.

**Comment by electromuis** [comment]
@coderjo 
The daemon clears the folder when an unknown card is scanned (for EC we used empty cards and marked them: Check out)
Also, when someone is checked and then someone else scans the card it removes the existing folder, and loads the new profile.

The goal is to also check out when the same card is scanned twice, but this is not implemented as of speaking.

Windows is supported! (At least it worked like this on my w10 machine) But for now, only 1 reader can be used. Using multiple is on the schedule though. 

As for copyright, i ll change that with my next update. Also want to notify stepmania when the last modification date changes

**Comment by electromuis** [comment]
Updated the copyright description,
Also added a lua function that allows song filtering from the theme. In my case i use it to hide certain packs in a Blacklist.txt. The packs become shown when a keycode is used

File: src/NotesWriterSM.cpp [hunk 1]
```diff
@@ -376,4 +381,4 @@ bool NotesWriterSM::WriteEditFileToMachine( const Song *pSong, Steps *pSteps, RS
  * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
Comment (shakesoda) [inline src/NotesWriterSM.cpp]: please keep newlines at end, not having them causes warnings.

File: src/NotesWriterSM.h [hunk 2]
```diff
@@ -3,45 +3,47 @@
...
-	 * @param pSteps the Steps in question.
-	 * @param sErrorOut any error messages that may have occurred.
-	 * @return its success or failure. */
-	bool WriteEditFileToMachine( const Song *pSong, Steps *pSteps, RString &sErrorOut );
+    /**
+     * @brief Write the song out to a file.
+     * @param sPath the path to write the file.
+     * @param out the Song to be written out.
+     * @return its success or failure. */
+    bool Write( RageFileBasic &file, Song &out, const vector<Steps*>& vpStepsToSave );
```
Comment (shakesoda) [inline src/NotesWriterSM.h]: fix style, we indent with tabs

File: src/NotesWriterSM.h [hunk 3]
```diff
@@ -61,4 +63,4 @@ namespace NotesWriterSM
  * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
Comment (shakesoda) [inline src/NotesWriterSM.h]: add newline

File: src/StatsManager.cpp [hunk 4]
```diff
@@ -227,60 +234,237 @@ void StatsManager::CommitStatsToProfiles( const StageStats *pSS )
...
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_HOLDROLLS, opts.m_bTransforms );
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_NOJUMPS, opts.m_bTransforms );
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_NOHANDS, opts.m_bTransforms );
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_NOLIFTS, opts.m_bTransforms );
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_NOFAKES, opts.m_bTransforms );
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_NOQUADS, opts.m_bTransforms );
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_NOSTRETCH, opts.m_bTransforms );
+
+#define ADD_FLOAT_OPTION( parent, name, opts ) \
+    do { \
```
Comment (shakesoda) [inline src/StatsManager.cpp]: space -> tab

File: src/StatsManager.cpp [hunk 5]
```diff
@@ -458,4 +642,4 @@ LUA_REGISTER_CLASS( StatsManager )
  * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
Comment (shakesoda) [inline src/StatsManager.cpp]: add newline

File: src/StatsManager.h [hunk 6]
```diff
@@ -7,36 +7,40 @@
 class StatsManager
 {
 public:
-	StatsManager();
-	~StatsManager();
+    StatsManager();
```
Comment (shakesoda) [inline src/StatsManager.h]: change to tab

File: src/StatsManager.h [hunk 7]
```diff
@@ -68,4 +72,4 @@ extern StatsManager*	STATSMAN;	// global and accessible from anywhere in our pro
  * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
Comment (shakesoda) [inline src/StatsManager.h]: add newline

File: src/UnlockManager.cpp [hunk 8]
```diff
@@ -412,6 +412,42 @@ UnlockEntryStatus UnlockEntry::GetUnlockEntryStatus() const
...
+
+	if(songEnabled.GetLuaType() != LUA_TFUNCTION)
+	{
+		LuaHelpers::ReportScriptErrorFmt("SongEnabledFunction metric for UnlockManager must be a function.");
+		return true;
+	}
+
+	Lua *L = LUA->Get();
+	songEnabled.PushSelf(L);
+    pSong->PushSelf(L);
```
Comment (shakesoda) [inline src/UnlockManager.cpp]: change to tabs

File: src/arch/MemoryCard/MemoryCardDriver.cpp [hunk 9]
```diff
@@ -7,6 +7,23 @@
...
 
+enum MemoryCardDriverType {
+	MemoryCardDriverType_Usb,
+	MemoryCardDriverType_Directory,
+	NUM_MemoryCardDriverType,
+	MemoryCardDriverType_Invalid
+};
+
+static const char *MemoryCardDriverTypeNames[] = {
+		"USB",
```
Comment (shakesoda) [inline src/arch/MemoryCard/MemoryCardDriver.cpp]: double indented

File: src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    if( info.st_mode & S_IFDIR ) {
+        return true;
+    }
+
+    return false;
+}
+
+bool MemoryCardDriverThreaded_Folder::TestWrite( UsbStorageDevice* pDevice )
+{
+	//TODO
```
Comment (shakesoda) [inline src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp]: I won't block the merge on this but it'd be nice if this todo gets sorted out in the near future :)

File: src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+	return true;
+}
+
+int MemoryCardDriverThreaded_Folder::GetActivePlayers()
+{
+    int ret = 0;
+
+    FOREACH_PlayerNumber( p )
```
Comment (shakesoda) [inline src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp]: change to tabs

File: src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+bool MemoryCardDriverThreaded_Folder::TestWrite( UsbStorageDevice* pDevice )
+{
+	//TODO
+
+	return true;
+}
+
+int MemoryCardDriverThreaded_Folder::GetActivePlayers()
```
Comment (shakesoda) [inline src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp]: given that this function returns a bitmask, it should probably be renamed to GetActivePlayerMask()

File: src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+{
+    return GetActivePlayers() != m_LastDevices;
+}
+
+void MemoryCardDriverThreaded_Folder::GetUSBStorageDevices( vector<UsbStorageDevice>& vDevicesOut )
+{
+	LOG->Trace( "GetUSBStorageDevices" );
+
+	vDevicesOut.clear();
+    m_LastDevices = GetActivePlayers();
```
Comment (shakesoda) [inline src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp]: change to tabs

**Comment by coderjo** [comment]
I'd suggest putting the SongEnabled commit in a separate PR, as it is a separate feature from support for padmiss.

**Comment by electromuis** [comment]
Sorry for the late response, appllied codestyle fixes. Also removed the song filter function to put in a seperate request

**Comment by shakesoda** [comment]
thanks, looks good to me 👍

**Comment by electromuis** [comment]
It shouldn't have, first i rolled back 1 commit. Then did the fixes

**Pull request merged**